### DISCUSSION TITLE
Are there any cross-platform GUI frameworks here which work out-of-the-box?

### DISCUSSION THREAD

**Discussion by SamuelMarks** [microsoft/vcpkg#18956] [Q&A]
I've recently been trying to get an open-source cross-platform GUI library—in C or C++—working with an open-source cross-platform C/C++ package manager.

Experience here with [`vcpkg`](https://github.com/microsoft/vcpkg):

  - [nanogui](https://github.com/mitsuba-renderer/nanogui) @ https://github.com/microsoft/vcpkg/issues/17052
  - [fltk](https://github.com/fltk/fltk) @ https://github.com/microsoft/vcpkg/issues/17053
  - [libui](https://github.com/andlabs/libui) @ https://github.com/microsoft/vcpkg/issues/17054
  - [nuklear](https://github.com/Immediate-Mode-UI/Nuklear) @ https://github.com/microsoft/vcpkg/issues/17055
  - [tgui](https://github.com/texus/TGUI) @ https://github.com/microsoft/vcpkg/issues/17056
  - [imgui](https://github.com/ocornut/imgui) @ https://github.com/microsoft/vcpkg/issues/17058
  - [glui](https://github.com/libglui/glui) @ https://github.com/microsoft/vcpkg/issues/17059
  - [wxwidgets](https://github.com/wxWidgets/wxWidgets) @ https://github.com/microsoft/vcpkg/issues/17072
  - [allegro5](https://github.com/liballeg/allegro5) @ https://github.com/microsoft/vcpkg/issues/17108
  - [gtk](https://www.gtk.org), [gtkmm](https://www.gtkmm.org) @ https://github.com/microsoft/vcpkg/issues/17216
  - [nana](https://github.com/cnjinhao/nana) @ https://github.com/microsoft/vcpkg/issues/17291
  - [dlib](https://github.com/davisking/dlib) @ ([LAPACK](http://www.netlib.org/lapack/): https://github.com/microsoft/vcpkg/issues/17293, GUI not launching: https://github.com/microsoft/vcpkg/issues/17295)

Experience with [`hunter`](https://github.com/cpp-pm/hunter):

  - [imgui](https://github.com/ocornut/imgui) (evolved into missing [allegro5](https://github.com/liballeg/allegro5)) @ https://github.com/cpp-pm/hunter/issues/376

Experience with [`sw`](https://github.com/SoftwareNetwork/sw):

  - [imgui](https://github.com/ocornut/imgui) @ https://github.com/SoftwareNetwork/sw/issues/64

Experience with [`conan`](https://github.com/conan-io/conan):

  - [imgui](https://github.com/ocornut/imgui) and [nuklear](https://github.com/Immediate-Mode-UI/Nuklear) (really [allegro5](https://github.com/liballeg/allegro5)) @ https://github.com/conan-io/conan/issues/8763
  - [wxwidgets](https://github.com/wxWidgets/wxWidgets) @ https://github.com/conan-io/conan/issues/8782
  - [libui](https://github.com/andlabs/libui) @ https://github.com/conan-io/conan/issues/8783
  - [tgui](https://github.com/texus/TGUI) @ https://github.com/conan-io/conan/issues/8784

Experience with [`conda`](https://github.com/conda/conda):

  - [fltk](https://github.com/fltk/fltk) @ https://github.com/conda/conda/issues/10601
  - [wxwidgets](https://github.com/wxWidgets/wxWidgets) @ https://github.com/conda/conda/issues/10602

---

You folks have been helpful and are commenting—and sometimes providing patches—for most of my issues here. But I still don't have a working 'Hello World' from any cross-platform C/C++ GUI library. Any suggestions?

**Comment by Neumann-A** [comment]
Qt should work (using CMake). At least the cmake test port and paraview work on windows & linux.

**Comment by SamuelMarks** [comment]
@Neumann-A My only issue with Qt—and the reason I haven't tested it in the above list—is the licensing issue. Commercial vs noncommercial usage.

Also for my purposes I don't need anything too fancy, just forms, a tree view, [stretch goal: image and video rendering], and maybe a graph or two… I even tried doing it myself (link with Windows SDK to create windows as they recommend, but [linking with macOS failed to show anything on screen](https://github.com/microsoft/vcpkg/pull/17095#issuecomment-813249549)).

I was rather surprised that none of the 9 vcpkg packaged UI frameworks worked out-of-the-box. Hopefully as those issues progress someone will contribute enough PRs that one works (there are some open PRs now; hopefully will be completed). Might need to send a few more myself…

**Comment by Neumann-A** [comment]
Qt is the only GUI Framework with test ports if it actually works. It has cmake in scripts/test_ports and paraview. So maybe vcpkg needs some test_ports for the other gui frameworks.

**Comment by NicolBolas** [comment]
@SamuelMarks 

> I was rather surprised that none of the 9 vcpkg packaged UI frameworks worked out-of-the-box. 

I'm not sure what you mean by "out-of-the-box". I had no trouble installing wxWidgets through vcpkg and finding a ["Hello World" tutorial](https://docs.wxwidgets.org/trunk/overview_helloworld.html) that could work with it (thus proving that vcpkg installed it correctly). Vcpkg installs libraries; it doesn't install applications. Libraries can be used to build applications, but you have to bring the application code yourself.

So they do work "out-of-the-box"; they just don't write the application code for you. Nor should they. So I don't think the problem is with vcpkg so much as what you expect from it.

**Comment by SamuelMarks** [comment]
@NicolBolas I also tried wxwidgets, see https://github.com/microsoft/vcpkg/issues/17072

Hello World didn't work. YMMV.
